inverted_index.py

```diff
@@ -0,0 +1,354 @@
+"""
+class InvertedIndex:
+
+    BLOCK_SIZE_LIMIT = 100000
+
+    def __init__(self, in_dir = "", out_dict = "dictionary.txt", out_postings = "postings.txt"):
+        print("Initialise Inverted Indexes...")
+
+        self.in_dir = in_dir
+        self.out_dict = out_dict
+        self.out_postings = out_postings
+
+        # Data Structures to be read for creating files
+        # To be used for creating indexes
+        self.postings = defaultdict(list) # key: Term, Value: List of doc_id
+        self.dictionary = set() # Terms
+
+        # line_offset list tells us the offset value for each line in the posting file
+        # To be used for fulfilling search queries
+        self.line_offset = []
+
+        # Load Dictionary Terms into memory when search.py initialises inverted_index
+        if in_dir == "":
+            self.LoadDictionaryFromMem()
+
+    """
+        ////////////////////////////////////////
+        Methods for index.py
+        ////////////////////////////////////////
+    """
+
+    """
+           Method to read the data file and fill up self.postings and self.dictionary.
+           
+           Current implementation only removes punctuations, case folding and do word stemming.
+           It does not remove stop words.   
+    """
+    def ConstructIndex(self):
+        print("Constructing Indexes...")
+
+        stemmer = PorterStemmer()
+        translator = str.maketrans('','', string.punctuation)
+
+        block_index = 0
+        curr_block_size = 0
+
+        self.ResetFiles()
+
+        # Get all the file names in reuters
+        all_files = []
+        for doc_id in os.listdir(self.in_dir):
+            all_files.append(int(doc_id))
+
+        # Read in ascending order of their file names
+        for doc_id in sorted(all_files):
+            for line in self.ReadFromFile(self.in_dir + "/" + str(doc_id)):
+                # Tokenize by sentences
+                for s_token in sent_tokenize(line):
+                    # Tokenize by word
+                    for w_token in word_tokenize(s_token):
+
+                        # Remove Punctuation
+                        w_token = w_token.translate(translator)
+
+                        # Word Stemming & Case-Folding
+                        term = stemmer.stem(w_token).lower()
+
+                        # Remove numbers
+                        if w_token.isnumeric() or len(term) == 0:
+                            continue
+
+                        if term not in self.dictionary:
+                            self.dictionary.add(term)
+                        if str(doc_id) not in self.postings[term]:
+                            self.postings[term].append(str(doc_id))
+
+                        curr_block_size += 1
+
+                        # Write the previous items to new block
+                        if curr_block_size == self.BLOCK_SIZE_LIMIT:
+                            block_index += 1
```

Comment:
maybe cleaner to increment block_index after writing, so that later on we don't have to skip block 0